src/rust/collections/concurrent_ring.rs

```diff
@@ -0,0 +1,509 @@
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// A lock-free, multi-reader, multi-writer, variable-sized messaging buffer. It uses a [push_offset] to indicate where
+/// to write incoming buffers and a [pop_offset] for where to read incoming buffers. Additionally each buffer contains
+/// a [HEADER_SIZE] header that indicates the length of the data (and as a result, pointing to the next valid buffer in
+/// the ring). This header also serves as a locking mechanism: on push, the writer atomically sets the header to
+/// indicated valid data that is ready to be read; on pop, the reader atomically sets the header to 0 in order to
+/// "claim"/lock the buffer from other concurrent readers.
+///
+/// For correctness, the following invariants
+/// must hold:
+/// 0. push_offset and pop_offset must monotomically increase, except when wrapping around
+/// 1. push_offset == pop_offset only if queue is empty.
+/// 2. push_offset == pop_offset - HEADER_SIZE only if queue is full. We only utilize capacity - HEADER_SIZE of the
+/// ring buffer so that we can distinguish between these two states.
+/// 3. First 16 bytes of every buffer is a header indicating the length of the buffer.
+/// 4. If the length header of a buffer is non-zero, the data in the buffer must be valid.
+/// 5. If the buffer is valid, it must exist between the push_offset and pop_offset. As a result, the pop_offset can
+///    never overtake the push_offset.
+/// 6. If the length header is zero, the data in the buffer may be valid but if it is valid, it is locked and should
+/// not be modified or read.
+/// 7. If the 16 bytes pointed to by pop_offset are zero, then there is either another ongoing pop or there is no valid
+/// data in the buffer.
+pub struct ConcurrentRingBuffer {
+    // Indexes the first empty byte where buffers can be enqueued.
+    push_offset: *mut usize,
+    // Indexes the first buffer that can be popped.
+    pop_offset: *mut usize,
+    // Underlying buffer.
+    buffer: raw_array::RawArray<u8>,
+    /// Is the underlying memory managed by this module?
+    is_managed: bool,
+}
+
+//======================================================================================================================
+// Associated Functions
+//======================================================================================================================
+
+/// Associated functions.
+impl ConcurrentRingBuffer {
+    /// Creates a ring buffer.
+    #[allow(unused)]
+    pub fn new(capacity: usize) -> Result<Self, Fail> {
+        // Check if capacity is invalid.
+        let layout: Layout = Layout::new::<usize>();
+
+        let push_offset: *mut usize = unsafe {
+            let ptr: *mut usize = alloc::alloc(layout) as *mut usize;
```

Comment:
Can this block be extracted into a separate function? It's also repeated below.

Reply:
Sure, but this block is the C's equivalent for:

```c
unsinged *ptr = malloc(sizeof(unsigned));
if (ptr == NULL) {
    exit(1);
}
```

Thus you are asking for a `alloc_unsigned()` function. Is that it?

Reply:
Done.